src/count_loc.jl

```diff
@@ -69,5 +83,184 @@ end
 count_readme(table) = count_docs(table, ("readme", "readme.md"))
 
 count_julia_loc(pkg::Package, args...) = count_julia_loc(pkg.lines_of_code, args...)
+count_docstrings(pkg::Package, args...) = count_docstrings(pkg.lines_of_code, args...)
 count_docs(pkg::Package, args...) = count_docs(pkg.lines_of_code, args...)
 count_readme(pkg::Package, args...) = count_readme(pkg.lines_of_code, args...)
+
+#####
+##### Custom Julia line counting
+#####
+
+# We can't do this at the level of SyntaxNode's because we've lost whitespace & comments already
+# So we will go to GreenNode's
+
+# Avoid piracy by defining AbstractTrees definitions on a wrapper
+struct GreenNodeWrapper
+    node::JuliaSyntax.GreenNode
+    source::JuliaSyntax.SourceFile
+end
+
+function AbstractTrees.children(wrapper::GreenNodeWrapper)
+    return map(n -> GreenNodeWrapper(n, wrapper.source), JuliaSyntax.children(wrapper.node))
+end
+
+function parse_green_one(file_path)
+    file = read(file_path, String)
+    @debug(string("Parsing ", file_path))
+    parsed = JuliaSyntax.parse(JuliaSyntax.GreenNode, file; ignore_trivia=false)
+    return GreenNodeWrapper(parsed[1], JuliaSyntax.SourceFile(file; filename=basename(file_path)))
+end
+
+# Module to make it easier w/r/t/ import clashes
+module CategorizeLines
+
+using JuliaSyntax: GreenNode, is_trivia, haschildren, is_error, children, span, SourceFile, source_location
+
+struct NodeSummary
+    starting_line::Int
+    ending_line::Int
+    summary::String
+end
+
+# Based on the recursive printing code for GreenNode's
+# Here, instead of printing, we collect all of the objects
+# that start on a given line.
+function categorize_lines!(d, node, source, nesting=0, pos=1)
+    starting_line, _ = source_location(source, pos)
+    v = get!(Vector{NodeSummary}, d, starting_line)
+
+    is_leaf = !haschildren(node)
+    if !is_leaf
+        new_nesting = nesting + 1
+        p = pos
+        for x in children(node)
+            categorize_lines!(d, x, source, new_nesting, p)
+            p += x.span
+        end
+        ending_line, _ = source_location(source, p)
+    else
+        ending_line = starting_line
+    end
+    push!(v, NodeSummary(starting_line, ending_line, summary(node),
+    # is_error=is_error(node), is_leaf, is_trivia=is_trivia(node), nesting
+    ))
+    return nothing
+end
+
+export NodeSummary
+end
+
+using .CategorizeLines
+
+# TODO:
+# Handle `@doc` calls?
+# What about inline comments #= comment =#?
+# Can a docstring not start at the beginning of a line?
```

Comment:
Yes.

```
julia> JuliaSyntax.parseall(JuliaSyntax.SyntaxNode, "x; \"some docstring\"\nf")
line:col│ tree                                   │ file_name
   1:1  │[toplevel]
   1:1  │  [toplevel]
   1:1  │    x
   1:3  │    [macrocall]
   1:3  │      core_@doc
   1:4  │      [string]
   1:5  │        "some docstring"
   2:1  │      f
```

Reply:
ok, that makes sense, thanks for the example. For the purpose "counting lines of code", I think I need to make a decision here if this line counts as a "docstring" line or a "code" line. My inclination is to just treat it as a code line, since that seems easier implementation-wise and I think it is also fairly explainable (it counts as the first thing that happens on that line).

Reply:
It's a pretty unusual case to be honest, so I don't actually think it matters for the purposes of package stats.

But you could count it as both a code line and a docstring line if you wanted? It is, after all, literally both.

Depends if you think it's important to have a 1:1 mapping between lines and classification

Reply:
I think it is important, yeah. I think it “feels” buggy otherwise, if you try it on a file and the counts don’t add up to the line count of the file you see in your editor. And I think it adds extra conceptual complexity.